### PR TITLE
feat: handle retries for conditionally idempotent bucket ops

### PR DESCRIPTION
**Pull request by ddelgrosso1**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-storage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: src/bucket.ts [hunk 1]
```diff
@@ -616,6 +621,8 @@ class Bucket extends ServiceObject {
   getFilesStream: Function;
   signer?: URLSigner;
 
+  originalRetryValue?: boolean;
```
Comment (shaffeeullah) [inline src/bucket.ts:624]: Maybe rename this to `instanceRetryValue`? Makes it more apparent that we want to override the value set at the Storage constructor level.

File: src/bucket.ts [hunk 2]
```diff
@@ -264,6 +264,11 @@ export enum BucketActionToHTTPMethod {
   list = 'GET',
 }
 
+export enum Methods {
```
Comment (shaffeeullah) [inline src/bucket.ts:267]: The name `Methods` seems vague/confusing to me if I was picking this up knowing nothing, but I'm not sure what a better option would be.

File: src/bucket.ts [hunk 3]
```diff
@@ -3951,6 +4051,25 @@ class Bucket extends ServiceObject {
   getId(): string {
     return this.id!;
   }
+
+  disableAutoRetryConditionallyIdempotent_(
+    coreOpts: any,
```
Comment (shaffeeullah) [inline src/bucket.ts:4056]: Is `any` the best option here? Can it be object?

File: src/bucket.ts [hunk 4]
```diff
@@ -3951,6 +4051,25 @@ class Bucket extends ServiceObject {
   getId(): string {
     return this.id!;
   }
+
+  disableAutoRetryConditionallyIdempotent_(
```
Comment (shaffeeullah) [inline src/bucket.ts:4055]: This function is really cool. Thanks for doing this!

File: src/bucket.ts [hunk 5]
```diff
@@ -3814,6 +3908,12 @@ class Bucket extends ServiceObject {
       options
     );
 
+    // Do not retry if precondition option ifMetagenerationMatch is not set
+    let numberOfRetries = this.storage.retryOptions.maxRetries;
```
Comment (shaffeeullah) [inline src/bucket.ts:3912]: Do we want to name this `maxRetries` for consistency? Or is it confusing to use the same name again? I guess it's not hard to understand, but sometimes fewer than this number of retries will occur if it succeeds on the first retry (for example)

**Review by JustinBeckwith — COMMENTED** [review]
Could we add tests?

File: src/bucket.ts [hunk 6]
```diff
@@ -1233,8 +1242,18 @@ class Bucket extends ServiceObject {
       return apiFormattedRule;
     });
 
+    this.disableAutoRetryConditionallyIdempotent_(
```
Comment (tritone) [inline src/bucket.ts:1245]: I'm not any kind of node expert but it looks like you are disabling the retry during the setMetadata operation and then reenabling it afterwards. If so, does that pose any kind of concurrency/race condition issues with the client? Also, will user configuration choices override this logic?

**Review by tritone — APPROVED** [review]

File: src/bucket.ts (hunk 6 above)
Comment (ddelgrosso1) [reply src/bucket.ts:1245]: There shouldn't be an issue with concurrency but I will make sure there isn't the possibility of a race condition. This is overriding the users configuration temporarily so actually the opposite of your second question.

File: src/bucket.ts (hunk 3 above)
Comment (ddelgrosso1) [reply src/bucket.ts:4046]: Object won't work, I suppose I could mimic / clone the type from common if that is preferred?

File: src/bucket.ts (hunk 6 above)
Comment (tritone) [reply src/bucket.ts:1245]: Gotcha. Does that mean that users can't configure the library to retry in spite of the operation being non-idempotent?

Comment (ddelgrosso1) [reply src/bucket.ts:1245]: The user can still configure the library to retry. However, for these conditionally idempotent operations if the correct precondition is not set, the library will override the user setting and turn off retries which is what the [spec](https://cloud.google.com/storage/docs/retry-strategy) notes.

File: src/bucket.ts (hunk 3 above)
Comment (shaffeeullah) [reply src/bucket.ts:4046]: I think either that or disabling linting on this line. Object from common is what we've done in the past, so might be a good idea. It creates duplicate code though :( 

Your call on what you think is best.

File: test/bucket.ts [hunk 7]
```diff
@@ -1532,6 +1581,16 @@ describe('Bucket', () => {
         done();
       });
     });
+
+    it('should disable autoRetry when ifMetagenerationMatch is undefined', done => {
+      bucket.setMetadata = () => {
+        assert.strictEqual(bucket.storage.retryOptions.autoRetry, false);
+      };
+
+      bucket.enableLogging({prefix: PREFIX}, (err: Error | null) => {
```
Comment (shaffeeullah) [inline test/bucket.ts:1590]: can you remove `err` here to get rid of the linter warning?

File: test/bucket.ts [hunk 8]
```diff
@@ -2131,6 +2199,14 @@ describe('Bucket', () => {
         done();
       });
     });
+
+    it('should disable autoRetry when ifMetagenerationMatch is undefined', done => {
+      bucket.setMetadata = (metadata: {}, options_: SetFileMetadataOptions) => {
```
Comment (shaffeeullah) [inline test/bucket.ts:2204]: linter warnings also on `metadata` and `options`

File: test/bucket.ts [hunk 9]
```diff
@@ -2553,6 +2681,18 @@ describe('Bucket', () => {
       });
     });
 
+    it('should disable autoRetry when ifMetagenerationMatch is undefined', done => {
+      const fakeFile = new FakeFile(bucket, 'file-name');
+      fakeFile.isSameFile = () => {
+        return true;
+      };
+      const options = {destination: fakeFile, metadata};
+      bucket.upload(filepath, options, (err: Error, file: FakeFile) => {
```
Comment (shaffeeullah) [inline test/bucket.ts:2690]: linter warnings on `err` and `file`

**Review by shaffeeullah — APPROVED** [review]
Thanks, Denis!!

**Pull request merged**